include/readout/RawWIBTp.hpp

```diff
@@ -0,0 +1,235 @@
+              << " fiber:" << unsigned(fiber_no) << " crate:" << unsigned(crate_no)
+              << " timestamp:" << timestamp()
+              << '\n';
+  }
+
+  void printHex() const
+  {
+    std::cout << "Printing raw WIB TP header:\n";
+    std::cout << std::hex << "flags:" << flags << " slot:" << slot_no << " wire:" << wire_no
+              << " fiber:" << fiber_no << " crate:" << crate_no << " timestamp:" << timestamp() << std::dec << '\n';
+  }
+
+  void printBits() const
+  {
+    std::cout << "flags:" << std::bitset<13>(flags)
+              << " slot:" << std::bitset<3>(slot_no) << " wire:" << std::bitset<8>(wire_no)
+              << " fiber:" << std::bitset<3>(fiber_no) << " crate:" << std::bitset<5>(crate_no) 
+              << " timestamp: " << timestamp() << '\n';
+  }
+};
+
+//========================
+// TP data struct
+//========================
+struct TpData
+{
+  // This struct contains three words of TP values that form the main repeating
+  // pattern in the TP block.
+  word_t end_time: 16, start_time: 16;
+  word_t peak_time: 16, peak_adc: 16;
+  word_t hit_continue: 1, flags: 15, sum_adc: 16;
+
+  void print() const
+  {
+    std::cout << "end_time:" << unsigned(end_time) << " start_time:" << unsigned(start_time)
+              << " peak_time:" << unsigned(peak_time) << " peak_adc:" << unsigned(peak_adc)
+              << " hit_continue:" << unsigned(hit_continue) << " flags:" << unsigned(flags) 
+              << " sum_adc:" << unsigned(sum_adc) 
+              << '\n';
+  }
+
+  void printHex(const int& i) const
+  {
+    std::cout << "Printing raw WIB TP " << i << ":\n";
+    std::cout << std::hex << "end_time:" << end_time << " start_time:" << start_time 
+              << " peak_time:" << peak_time << " peak_adc:" << peak_adc 
+              << " hit_continue:" << hit_continue << " flags:" << flags 
+              << " sum_adc:" << sum_adc << std::dec << '\n';
+  }
+};
+
+//========================
+// TP pedestal information struct
+//========================
+struct TpPedinfo
+{
+  // This struct contains three words: one carrying median and accumulator and two padding words
+  word_t accumulator: 16, median: 16;
+  word_t padding_2: 16, padding_1: 16;
+  word_t padding_4: 16, padding_3: 16;
+
+  void print() const 
+  {
+    std::cout << "median:" << unsigned(median) << " accumulator:" << unsigned(accumulator)
+              << '\n';
+  }
+
+  void printHex() const
+  {
+    std::cout << "Printing raw WIB TP pedinfo:\n";
+    std::cout << std::hex << "median:" << median << " accumulator:" << accumulator 
+              << " padding_1:" << padding_1 << " padding_2:" << padding_2 
+              << " padding_3: "<< padding_3 << " padding_4:" << padding_4 << std::dec << '\n';
+  }
+};
+
+//========================
+// TpData block
+//========================
+struct TpDataBlock
+{
+  TpData tp;
+
+  std::vector<TpData> block;
```

Comment:
Using a `std::vector` data member here means that the contents of the `RawWIBTp` frame won't be contiguous in memory, and so you won't be able to use the following idiom:

```cpp
char* buffer=get_a_frame_from_the_hardware();
RawWIBTp* frame=reinterpret_cast<RawWIBTp*>(buffer);
frame->get_tp(3).end_time;
```

To put it another way: a frame that comes from the firmware will consist of a single, contiguous block of memory (right?). You have to be able to interpret that block of memory. In the "overlay" approach -- which is what I think is intended here -- you have to make the memory-layout of your classes match the memory-layout of the block of memory that comes from the firmware.

Reply:
I understand better now, indeed we should be able to use 'frame' as outlined in the snippet above but I haven't fully realised it up to now.Yes, this is completely right. At the moment, when we are reading the data from a binary input file this is exactly the case, i.e. there are no any header or trailer words that need to be removed.

Reply:
I will test that the above suggestion works. Before moving away from using a vector for storing the variable number of TPs, Roland suggested to perhaps have a discussion on udaq-software on slack but I see that Phil is not connected there.

Reply:
On the transition from firmware memory blocks to offline fragments and serialization I've read Phil's and Roland's comments to PR https://github.com/DUNE-DAQ/dataformats/pull/37 and cannot decide what to do at the moment.

Reply:
I strongly agree with @philiprodrigues . We can't have a vector as a member field in an "overlay". TBD after we move the rawTP to `dataformats`.